modules/azure/nsg.go

```diff
@@ -0,0 +1,194 @@
+package azure
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-09-01/network"
+)
+
+// GetDefaultNsgRulesClientE returns a rules client
+func GetDefaultNsgRulesClientE(subscriptionID string) (network.DefaultSecurityRulesClient, error) {
+	// Validate Azure subscription ID
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return network.DefaultSecurityRulesClient{}, err
+	}
+
+	nsgClient := network.NewDefaultSecurityRulesClient(subscriptionID)
+
+	// Get an authorizer
+	auth, err := NewAuthorizer()
+	if err != nil {
+		return network.DefaultSecurityRulesClient{}, err
+	}
+
+	nsgClient.Authorizer = *auth
+	return nsgClient, nil
+}
+
+// GetCustomNsgRulesClientE returns a rules client
+func GetCustomNsgRulesClientE(subscriptionID string) (network.SecurityRulesClient, error) {
+	// Validate Azure subscription ID
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return network.SecurityRulesClient{}, err
+	}
+
+	nsgClient := network.NewSecurityRulesClient(subscriptionID)
+
+	// Get an authorizer
+	auth, err := NewAuthorizer()
+	if err != nil {
+		return network.SecurityRulesClient{}, err
+	}
+
+	nsgClient.Authorizer = *auth
+	return nsgClient, nil
+}
+
+// GetAllNSGRulesE returns a slice containing all rules from the NSG (including defauls and custom)
+func GetAllNSGRulesE(resourceGroupName, nsgName, subscriptionID string) (NsgRuleSummaryList, error) {
+	defaultRulesClient, err := GetDefaultNsgRulesClientE(subscriptionID)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	customRulesClient, err := GetCustomNsgRulesClientE(subscriptionID)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	defaultRuleList, err := defaultRulesClient.ListComplete(context.Background(), resourceGroupName, nsgName)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	customRuleList, err := customRulesClient.ListComplete(context.Background(), resourceGroupName, nsgName)
+	if err != nil {
+		return NsgRuleSummaryList{}, err
+	}
+
+	rules := make([]NsgRuleSummary, 0)
+	bindRuleList(&rules, defaultRuleList)
+	bindRuleList(&rules, customRuleList)
```

Comment:
NIT: We try to avoid APIs that unnecessarily mutate data and stick with immutable/functional practices when possible. I think this would be clearer and more maintainable if `bindRules` returned new lists and those were then merged together:

```go
boundDefaultRules := bindRules(defaultRuleList)
boundCustomRules := bindRules(customRuleList)
allRules := append(boundDefaultRules, boundCustomRules...)
```

Reply:
Updated pattern to stop mutation and instead return new collection.